snuba/subscriptions/worker.py

```diff
@@ -2,47 +2,85 @@
 
 import itertools
 import json
-from concurrent.futures import Future, as_completed
-from typing import Mapping, NamedTuple, Optional, Sequence
+import time
+from concurrent.futures import Future, ThreadPoolExecutor, as_completed
+from typing import Mapping, NamedTuple, Optional, Sequence, Tuple
 
+from snuba.datasets.dataset import Dataset
+from snuba.request import Request
 from snuba.reader import Result
 from snuba.subscriptions.consumer import Tick
 from snuba.subscriptions.data import Subscription
-from snuba.subscriptions.executor import SubscriptionExecutor
 from snuba.subscriptions.scheduler import ScheduledTask, Scheduler
 from snuba.utils.codecs import Codec
+from snuba.utils.metrics.backends.abstract import MetricsBackend
+from snuba.utils.metrics.gauge import Gauge
+from snuba.utils.metrics.timer import Timer
 from snuba.utils.streams.batching import AbstractBatchWorker
 from snuba.utils.streams.kafka import KafkaPayload
 from snuba.utils.streams.producer import Producer
 from snuba.utils.streams.types import Message, Topic
+from snuba.web.query import parse_and_run_query
 
 
-class SubscriptionResultFuture(NamedTuple):
+class SubscriptionTaskResultFuture(NamedTuple):
     task: ScheduledTask[Subscription]
-    future: Future[Result]
+    future: Future[Tuple[Request, Result]]
```

Comment:
I'd rather the request and result were handled separately, but the request can't be generated before the future execution, otherwise it screws the timings all up because they include request validation, so if there is any queue delay that gets confusingly rolled into the query execution stats.